Strip the prefix from all paths when installing with DESTDIR
Otherwise we end up with <destdir>/<prefix>/<path>
--- meson-0.47.0/mesonbuild/scripts/__init__.py
+++ meson-0.47.0/mesonbuild/scripts/__init__.py
@@ -12,7 +12,18 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import os
+from ..mesonlib import is_windows
+
 def destdir_join(d1, d2):
+    # When the installation path starts with PREFIX, strip it away
+    if d1 and is_windows() and 'MSYSTEM' in os.environ:
+        assert os.path.isabs(d2) and 'MESON_INSTALL_PREFIX' in os.environ
+        prefix = os.environ['MESON_INSTALL_PREFIX']
+        rel_to_prefix = os.path.relpath(d2, prefix)
+        if not rel_to_prefix.startswith(os.pardir + os.sep):
+            return os.path.join(d1, rel_to_prefix)
+
     # c:\destdir + c:\prefix must produce c:\destdir\prefix
     if len(d1) > 1 and d1[1] == ':' \
             and len(d2) > 1 and d2[1] == ':':
--- meson-0.47.0/mesonbuild/minstall.py.orig
+++ meson-0.47.0/mesonbuild/minstall.py
@@ -286,6 +286,7 @@
         with open(datafilename, 'rb') as ifile:
             d = pickle.load(ifile)
         d.destdir = os.environ.get('DESTDIR', '')
+        os.environ['MESON_INSTALL_PREFIX'] = d.prefix
         d.fullprefix = destdir_join(d.destdir, d.prefix)
 
         if d.install_umask is not None:
